### DISCUSSION TITLE
Impossible to return multiple discounts from the same function

### DISCUSSION THREAD

**Discussion by mariusa** [Shopify/function-examples#194] [Feedback: Discounts APIs]
There's a huge limitation on functions (discounts) which makes this unusable in real life cases: even though a function can return a list of multiple discounts (say, one for each cart line item), Shopify will apply only one of it. We need all the discounts to be applied.

For example: Volume discounts have different tiers for different quantities. A cart line could have quantity 2 and qualify for 1st tier (10% discount), another cart line could have quantity 6 and qualify for 2nd tier (20% discount). This is impossible to handle now, as even though the function returns 2 different discounts to apply, Shopify applies only one.
Need  DiscountApplicationStrategy: ALL, not just one: https://shopify.dev/api/functions/reference/product-discounts/graphql/common-objects/discountapplicationstrategy

There are forum posts about this, but nobody at Shopify didn't even reply:
https://community.shopify.com/c/shopify-functions/shopify-functions-product-discount-api-multiple-discounts/td-p/1693369
https://community.shopify.com/c/shopify-functions/need-discountapplicationstrategy-all/td-p/1865853

**Comment by rahultiwari-hulkapps** [comment]
+1
That is what we are looking for. We really need DiscountApplicationStrategy: ALL so different percent/fixed discounts can be applied multiple line targets.

**Comment by rahultiwari-hulkapps** [comment]
Hello Shopify, 

I hope you are looking here, is there any ETA for this? This is the main blocker for many developers.

Thank you so much.

**Comment by isque03** [comment]
Another +1
Ran into the same issue right away while working on implementing function discounts. Thanks for your help!

**Comment by mmarcon91** [comment]
We have the same problem

**Comment by GYMarshy** [comment]
+1 here too

**Comment by nfusionsolutions** [comment]
We need this also. Has Shopify support responded?

**Comment by UsmanFirdous** [comment]
I have same issue looking for solution

**Comment by sirowland** [comment]
Also looking for a solution to this.

Our use case:
- We have multiple cart line items that need their own specific discount
- we have custom products, so the product variant is not unique (we would have millions of variants).
 
If the target was a cart line id instead of a product variant id  and iterated through the cart lines, we would be able to run discounts like so:

cart line 1 (product variant 1 - width 1 length 2) -> function -> discount result 1 
cart line 2 (product variant 1 - width 2 length 3) -> function -> discount result 2 
cart line 3 (product variant 2 - width 1 length 2) -> function -> discount result 3

**Comment by cyberabis** [comment]
+1...

We might have a single volume discount setup for 2 products, say A and B.
Cart has x quantity of A, and y quantity of B. But the function is unable to apply 2 discounts across the cart line items.

**Comment by rahultiwari-hulkapps** [comment]
Any update/ETA here?

**Comment by mmarcon91** [comment]
> We’re removing the Script Editor app from the Shopify App Store on April 3, 2023

You decided to remove Shopify script from the store, which remained a possibility to solve this type of discounting. 
Now how can we solve for the new Shopify Plus this logic?

**Comment by rndgf** [comment]
+1

I join this waiting list :)
Just starting to play with functions, and want to implement some "private sales" discount rules (change a cart item price based on a variant metafield), seems to fit perfectly our need **EXCEPT THAT** it's impossible to have multiple discounts rules for the same function output... 

**_PLEASE_**, at least make us a statement ?

- Possible : Yes / No ?
- If Yes : an approximative ETA ?

I think it's a basic feature which need to be implement asap.
Currently, in my case, we need to rollback to "script editor" which is not the good path to 2024 end of life.

**Comment by ayushishah169** [comment]
I have the same issue with multi-tier discounts, looking for solution.

**Comment by Wtrapp** [comment]
+1

**Comment by gancena** [comment]
+1 many of our clients need/want this to work

**Comment by MurmeltierS** [comment]
It’s crazy that this is not an option yet - with Shopify Scripts now delisted this will get ugly soon

**Reply by mickeyambrosius** [reply]
Yeah I just found out that it's no longer possible to install the Script editor from the App Store. Been in a loop with Shopify support for 30 mins. Support trying to convince me it's better to switch to functions and me referring to this thread saying it's not possible yet. 

Anyone know if it's possible to install Script editor by a direct link?

**Reply by SiddharthSham** [reply]
It is possible to get a direct install link if your store is on a Plus plan, from my experience.

**Reply by adamkitain** [reply]
https://apps.shopify.com/script-editor

**Comment by valerii-stadnyk-moc** [comment]
+1

**Comment by realpratiknikam** [comment]
Based on Discounts AMA, this feature won't be available until end of year :( 

<img width="440" alt="image" src="https://user-images.githubusercontent.com/2853953/231815064-8fb55c34-096c-4f4b-8b9d-69515f4abfcc.png">

**Reply by jclay-pnw** [reply]
I was on that AMA as well and I didn't disagree with Adam when he posted that - since then we've gone back and re-evaluated and don't think this is as far off as that.

**Reply by wisniewski94** [reply]
Hi @jclay-pnw , can you share which Q of 23 you aim for :)? Knowing that recently we had a few minor updates on functions, most likely it's going to be Q3 - Q4?

**Reply by rahultiwari-hulkapps** [reply]
yes, we would like to know if there is any ETA for this.

**Comment by rahultiwari-hulkapps** [comment]
I am disappointed to know that Discount Team will address this issue later this year from the above screenshot, But I am still waiting for an official update from the Shopify team on this discussion. Hope they will make it a priority.

**Comment by adamkitain** [comment]
We 100% need this feature and it's blocking us from rolling functions out more generally.

**Comment by jclay-pnw** [comment]
Hi all - I'm the PM for discounts and appreciate all of the feedback on this particular issue. This is the highest priority on our roadmap for improving Discount Functions. I can't yet share specific timing on addressing this issue but it's actively being worked on. I will post more when I can share specific timelines!

**Reply by abhisheksingh75** [reply]
any updates here @jclay-pnw

**Comment by jesper-madsen** [comment]
Like every one else here, missing this feature keeps us away from migrating clients from Shopify Scripts. I'm sure there went a ton of thoughts into Discount Functions limitations, but this one is a odd limitation for sure.

**Comment by sethbeer** [comment]
The inability to apply a discount per cart line is a big problem and will block many of ours clients from making the switch at present. I like the idea of DiscountApplicationStrategy.All but for me it would be nicer to move the discount value to within the targets array. So instead of having a one price too many targets you have a one two one relationship instead. Something like the below.
`  return {
      discounts: [
        {
          targets: [
            {
              productVariant: {
                id
              },
              value: {
                percentage: {
                  value
                }
              }
            }
          ]
        }
      ],
      discountApplicationStrategy: DiscountApplicationStrategy.Maximum
    }`

**Comment by tarikhulk** [comment]
Are there any news on this?

**Comment by rnareset-dwhite** [comment]
It's been almost two months since we've heard from the PM that this is the highest priority. Would really like to know more. Is this going to be released before the Shopify Scripts deprecation, or are we going to have to come up with other solutions?

**Reply by SalTor** [reply]
After attending Shopify Office Hours for Checkout Extensibility and asking about this discounts issue I believe the latest on this front is waiting to get an update in their "Summer Editions" news release which might come in the next month or so

**Comment by AtulGargMT** [comment]
+1 from my side. It's a basic need to have different discounts on different items. We need this ASAP as this is blocking our Checkout migration from liquid based checkout to new checkout. Deprecation of liquid based checkout is also not that far now.

**Comment by pablogiralt** [comment]
+1

**Comment by BrandoCommando** [comment]
Hopefully the delay in API version 2023-07 is due to them trying to get this change in 🐎

**Comment by abhisheksingh75** [comment]
we are stuck because of this :(

**Comment by BrandoCommando** [comment]
So it seems like there are many people who are in favor of this concept, but not many are providing their actual use case. Shopify doesn't really care how many upvotes an issue gets if it doesn't understand the problem.
Here's an example:
- A shirt that sells for $10, pants that sell for $30, shoes that sell for $40, and a hat that sells for $20. Buy the shirt and pants together, get $10 off. Buy the hat and shows together, and get $20 off.

What happens when all 4 are in the cart together? The maximum would get a $20 discount, the "first" would get a $10 discount. Either way, customer service is getting a call. The function could group all the products together and give a $30 discount, but then it would discount everything by $15, which would then get messy when the customer wanted to return something.

DiscountApplicationStrategy may work decently for code-type discounts, but Shopify definitely needs to expand it for automatic discounts.

**Comment by ghost** [comment]
We all desperately need this feature. Since right now there's absolutely no way to discount multiple line items with differing discount values

**Comment by X4mp** [comment]
We are currently also blocked by lack of support to apply multiple discounts to multiple products on the cart.

**Comment by petarbeck** [comment]
We also need to apply a discount to all eligible line items and are only able to do so for the first item.

We are responding with DiscountApplicationStrategy.FIRST and multiple discounts, each referencing their target line item.
=> We expect every discount to be applied to their referenced targets.

This does not appear to be a limitation, but simply a bug - right?

Thanks

**Comment by pablogiralt** [comment]
@jclay-pnw do you have an update on this?

Shopify Discounts team should reconsider making the Script Editor app available again until this is addressed.  We have a new Plus project where we need this feature which we've implemented using Script lots of times and is not possible with Functions

**Reply by adamkitain** [reply]
You can still install the script editor here: https://apps.shopify.com/script-editor 🤫

**Reply by pablogiralt** [reply]
oh @adamkitain thanks so much, I thought it had been removed as a search does not return it

**Comment by overallduka** [comment]
+1 need it for my application OR need to remove the limitation of the maximum number of Automatic App Discounts, it seems that this is limited in every aspect.

**Comment by bsmithrainbowshops** [comment]
PLEASE, we atleast need a documented walkthrough as to how to apply this to run across a large cart of items. We routinely have customers with very large 100+ lineitems with all kinds of discounts that need to be applied to their cart. Shopify Scripts does this (slowly, occasionally times out), but it does so reliably. Please we need feature parity with what could be achieved today with Shopify Scripts.

**Reply by thosakwe** [reply]
Hey, I started another discussion specifically about running discount functions on large orders: https://github.com/Shopify/function-examples/discussions/329

I asked if the instruction limit for large orders can be raised, proportional to the number of cart lines. So if this is implemented, I think it would solve part of our problem.

**Comment by dianacheptene** [comment]
All, 
Thanks for bringing up / bumping this limitation on returning multiple discounts in discount functions. We’re currently working on enabling AllApplicable discounts to be returned.

We'll keep you in the loop with frequent updates on our progress.

**Reply by cyberabis** [reply]
Yes please! An approximate timeline will definitely help in managing expectations with merchants!

**Reply by jasonplumb** [reply]
+1 for an approximate timeline. It's evident that merchants provide the most valuable feedback w.r.t Shopify functions and checkout UI extensions so there should be a better dialog between Shopify internal and any public repo. 

Sunsetting the script editor and checkout.liquid combined with limited visibility is causing frustration at the best of times so an approximate timeline would help a lot and be beneficial for both parties.

@dianacheptene has progress been made in the last 3 weeks? Something like "we are at 1%, 5%, 90% to having an initial solution" is something I would personally like to see.

**Reply by amcaplan** [reply]
@Jay-Plumb also opened an issue, I'm closing as this is the correct place to have the conversation. https://github.com/Shopify/cli/issues/2709

**Reply by javl** [reply]
@dianacheptene  As Jay-Plumb said it would be great to have some sort of indication when this might become available. I just implemented a complicated setup dealing with multiple price breaks and quantities for very specific custom products, only to find out I can't apply the calculated discount to all products in the cart at the same time.

I'd love to know if I can wait for this to be updated, or need to find some workaround to use in the meantime.

**Reply by pablogiralt** [reply]
@dianacheptene it has been +1 month since your message, where you promised frequent updates.  Could you please give us a quick update on the status of this?

**Reply by dianacheptene** [reply]
Hi all, we're targeting November to release this update to all Discount Functions.

**Reply by adamkitain** [reply]
Will there be a developer preview version of it?

**Reply by peepers-rick** [reply]
@dianacheptene releasing, as in available to use in production immediately? I need this for a BFCM promotion and cutting it this close is making me exceptionally nervous.

**Comment by swat23** [comment]
+100 on this

**Comment by bluedge** [comment]
Much Needed bug Fix here. Merchants are furious!

**Comment by FrankBlueswitch** [comment]
can we get a timeline on this? we also have a client who does tiered pricing on a per product level depending on the qty of items in the cart. this was doable with the script editor app however since that is being removed next august they do not want to invest the time into setting up their new theme with this script. please some clarity on timelines would help us in deciding how to move forward.

**Reply by SalTor** [reply]
FWIW @FrankBlueswitch It’s checkout.liquid that’s going away August 2024 but not the Script Editor app. While it is going away too, it’s on a different and further away timeline that I don’t think has been disclosed

**Reply by wisniewski94** [reply]
@SalTor https://help.shopify.com/en/manual/checkout-settings/script-editor/migrating

On August 13, 2024, Shopify Scripts will be deprecated and will no longer be supported. Migrate your existing Scripts to [Shopify Functions](https://shopify.dev/api/functions) before this date.

**Reply by SalTor** [reply]
@wisniewski94 TY for this, I’m going to ask my Shopify account manager about this because this is different from what I’ve heard thus far

**Comment by gabrielvictoria96** [comment]
Hello,

I'm unsure if this situation applies to yours, but I've set up two automatic discounts using Shopify functions. These discounts are meant to be applied based on specific product attributes.

Like the print below, I've managed to apply different discounts to two products that possess distinct variant IDs. (each one it will be one discount applied).

However, a challenge I'm facing is that the discount target can only be the product variant, meaning the discount gets associated with the variant ID instead line ID. Consequently, if I have two products sharing the same variant ID but differing attributes, both products will receive the applied discount.

This issue is limitating me a lot. Merchants are requesting this particular feature, but unfortunately, there seems to be no feasible way to implement it correctly.

![image](https://github.com/Shopify/function-examples/assets/81654814/eb9df093-8b68-45ec-99eb-90025c4a7a27)

**Reply by overallduka** [reply]
Actually in my mind makes a lot of sense to apply the discount per line item key, instead of variant id(or both), since the discount is specifically for cart items.

**Comment by overallduka** [comment]
@gabrielvictoria96 is right, you can create up to 5 Automatic Discounts in Shopify, which means you can apply up to 5 of these discounts every time in the cart.

I've been testing this implementation of multiple discounts in the same cart, but it has a bug.

Let me explain the bug showing off:

![image](https://github.com/Shopify/function-examples/assets/4923856/1e4ed088-116e-4873-a92a-59ad9946896c)

I am using 2 equal separate Automatic Discounts here, both discounts are equal:

1x The Collection Snowboard: Oxygen + 1x The Collection Snowboard: Oxygen = 10% discount

The 2 outputs(1 for each Automatic Discounts) are equal:

> {
  "discounts": [
    {
      "message": "[PROD] Standard with Percentage OFF #1",
      "targets": [
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/45696112754985",
            "quantity": 1
          }
        },
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/45696112787753",
            "quantity": 1
          }
        }
      ],
      "value": {
        "percentage": {
          "value": 10
        }
      }
    }
  ],
  "discountApplicationStrategy": "FIRST"
}


However as seen in the image, just 1 discount is being entirely applied, for some reason the other discount is applying just in 1 product, this is a recurring error that I could reproduce other times.

**Reply by gabrielvictoria96** [reply]
Isn't because your quantity param?

**Reply by overallduka** [reply]
I don't think so, I have 4 products in the cart, 2 of each variant id, the discount should apply at quantities 1x1 and 1x1, it's applying 1x1 1x0 (???)

**Reply by gabrielvictoria96** [reply]
It's so weird... Did you try not sending the quantity param? Quantity isn't required (just the id of product variant).

**Reply by overallduka** [reply]
Well, the thing is: I need to discount just 1 quantity of each variant, if I remove the quantity param probably it will apply to all quantities of that variant which is not my goal. It's certainly a bug.

**Reply by gabrielvictoria96** [reply]
I can't reply the same bug, here is a scenario with two differents variants ID, applying the same discount once to each variant id and other scenario applying different discounts in different products. I send my quantity param equals 1.

![image](https://github.com/Shopify/function-examples/assets/81654814/0ed727d1-4511-43be-bfa8-0901971ea0f7)
![image](https://github.com/Shopify/function-examples/assets/81654814/726be6be-d569-4744-8b47-0aa3941031ee)

**Reply by andriusvo** [reply]
@gabrielvictoria96 Can you add your input and output of the function?

**Comment by rgilbert82** [comment]
Bump + 1. 
I also need to be able to apply different discount amounts on a line item basis. 

Here's an example with 3 line items: 
1. Line item 1 should be a 10% discount 
2. Line item 2 should be a 20% discount 
3. Line item 3 has the same variant ID as line item 1, but receives no discount.

**Comment by BrandoCommando** [comment]
I noticed the other day that "ALL" is now an acceptable value for `discountApplicationStrategy`. However, using this strategy with multiple discounts does not work. In my dev store I have a product discount function set up to give discounts to different pairs of items. In my example cart, I have 2 pairs of items that should receive a discount. Everything passes into the function correctly, and here's the output the function returns to Shopify:
```
{
  "discountApplicationStrategy": "ALL",
  "discounts": [
    {
      "message": "2-pack for $32",
      "value": {
        "fixedAmount": {
          "amount": 11.9,
          "appliesToEachItem": false
        }
      },
      "targets": [
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/44720327524665",
            "quantity": 1
          }
        },
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/44720326836537",
            "quantity": 1
          }
        }
      ]
    },
    {
      "message": "2-pack for $40",
      "value": {
        "fixedAmount": {
          "amount": 9.89,
          "appliesToEachItem": false
        }
      },
      "targets": [
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/44720321888569",
            "quantity": 1
          }
        },
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/44720323199289",
            "quantity": 1
          }
        }
      ]
    }
  ]
}
```

This results in only the first discount being applied:
<img width="440" alt="image" src="https://github.com/Shopify/function-examples/assets/97893/269f916b-972b-4a00-8a4c-dd3522a23e43">

Again, this is using the "ALL" discountApplicationStrategy, as opposed to "MAXIMUM". The [newly updated API reference for discountApplicationStrategy (must be "unstable")](https://shopify.dev/docs/api/functions/reference/product-discounts/graphql/common-objects/discountapplicationstrategy?api%5Bversion%5D=unstable) specifies "ALL: Apply all discounts with conditions that are satisfied." which would be great if it worked.

**Comment by milkmidi** [comment]
+1 here too

**Comment by mdriess** [comment]
We are also having to regularly explain to clients why they can't move xyz to Functions yet. Very excited for us to get this functionality! :) Any news / ETA? Thank you!!

**Comment by elvonkh** [comment]
+1
It would be good for partners to apply different discounts to a multiple line items with a single function.

**Comment by BrandoCommando** [comment]
Looks like this has been fixed, at least for my use cases! Thanks!

**Reply by rahultiwari-hulkapps** [reply]
In this case, can you share your sample toml of function? API version should be 'unstable' or '2024-01'?
Because for me DiscountApplicationStrategy::ALL is still not working?

**Reply by stark114** [reply]
Can you please your code bro? I need your help, please.

**Reply by BrandoCommando** [reply]
See my [previous post](#discussioncomment-7191358), same input, same output, same response, but now it "works". Unstable :: "ALL" got me there.

**Comment by AndreaMorone** [comment]
Stille not working, the output contains an array of different discounts targeting different products at 10% and 20% OFF, but it still discounts with the same amount. I think is still 'unstable' to use it.

<img width="890" alt="screen_1" src="https://github.com/Shopify/function-examples/assets/7823934/d196471d-ef08-4f90-a856-fd1fd9a81635">
<img width="412" alt="screen_2" src="https://github.com/Shopify/function-examples/assets/7823934/2b199492-fb2c-4539-80e3-01fe5b4d1f36">

**Reply by dianacheptene** [reply]
We're actively working on this and planning to release this feature in 2 weeks. I will post an update here so you can test it for your respective use cases.

**Reply by AndreaMorone** [reply]
Thank you, got it 👍🏻

**Reply by jacknavynicy** [reply]
Hi @dianacheptene will this release solved the issue of
like currently is pointing a product variant id, we have different discounts that will applied to the same product variant but different quantity, we want all discounts to be apply, need to apply to other product that do not have discount

or this only solve the issue of applying multiple discounts on same variant within in the same discount only?

**Reply by dostu** [reply]
I have observed the same behavior when testing. The discount amount is correct but the same message is used for all discounts.

**Reply by adamkitain** [reply]
Hey @dianacheptene - checking to see if you have a sense of the release for `DiscountApplicationStrategy.ALL` - should be soon, right? 🙏

**Reply by cyberabis** [reply]
Hi @dianacheptene, When is the GA release planned, we were hoping to have this before BFCM. Will this at least make it to 2024-01 release?

**Reply by RaphaelLimaDev** [reply]
Hello @dianacheptene, any news about it???

**Reply by Sanj718** [reply]
Hey guys, sorry for a bit out of topic questions, but how you been able to show custom discount message in checkout? I'm passing my message in target, but not seeing it in checkout...` "20% OFF VIP CUSTOMER"`

```javascript
  return {
    discounts: [
      {
        message: "20% OFF VIP CUSTOMER",
        targets: variantIds,
        value,
        
      },
    ],
  
    discountApplicationStrategy: DiscountApplicationStrategy.Maximum
  };
```

**Reply by AndreaMorone** [reply]
What API version are you using? ProductDiscount or OrderDiscount? The output seems correct to me. Does the discount apply without the message?

**Reply by Sanj718** [reply]
Yes, I've used 2023-10 and unstable, none of them works... it only shows Discount code as message.

**Reply by Sanj718** [reply]
I'm trying to implement Product Discount and custom message: 
![image](https://github.com/Shopify/function-examples/assets/7362855/91bce8b8-881f-4ddf-b6e2-582cf4e98f54)

Not sure why.... I was never able to see the custom message. Maybe there are some other pre-requisites I'm missing?

And this is my full output...
```javascript
{
  "discounts": [
    {
      "message": "20% OFF YOLO",
      "targets": [
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/42476793299132",
            "quantity": 2
          }
        }
      ],
      "value": {
        "percentage": {
          "value": 20
        }
      }
    }
  ],
  "discountApplicationStrategy": "ALL"
}
```

**Reply by Sanj718** [reply]
Any help here would be appreciated

**Comment by mhcdev** [comment]
> We're actively working on this and planning to release this feature in 2 weeks. I will post an update here so you can test it for your respective use cases.


Will this update allow us to select cart items by line item/key instead of variant id so we can target items properly for discounts? Or is this only going to allow a single function to return multiple discounts?

**Reply by AndreaMorone** [reply]
I think the second one, it’s about the discount application strategy, targets selector won’t change

**Reply by mhcdev** [reply]
I don't think that would be a satisfactory solution to this issue.  Most ppl need to be able to target specific line items for discounts not just variants in the cart.  Without that detail this discounting function is incomplete and will not be a proper replacement for scripts.

**Reply by javl** [reply]
Isn't it? Even though the target selector uses variant IDs, aren't those functionally the same as line items, as the script runs in a single cart at a time?

**Reply by nickwesselman** [reply]
You can follow the status of line item discount support under [this feature request](https://github.com/Shopify/function-examples/discussions/199).

**Reply by mhcdev** [reply]
Not really.  There is a chance a variant can be in the cart multiple times w/ different line item properties and need to have different discounts applied to each one. If we are only targeting by variantID we will never be able to apply discounts to the right items in this case.  This is why the item key was added to the REST API for cart line items.

**Reply by javl** [reply]
That makes sense, @mhcdev. Thanks.

**Comment by mishac** [comment]
You can have the same variant in different line items with different
discounts in the context of upsells, bundles, etc.

**Comment by mhcdev** [comment]
> You can follow the status of line item discount support under [this feature request](https://github.com/Shopify/function-examples/discussions/199).

I think they are pretty much the same issue. Both items need to be solved for functions to take over script discounts.

**Comment by sebastianpisula** [comment]
support for line item properties is very important IMHO

**Comment by dianacheptene** [comment]
All, this issue is specifically tracking the inability to return multiple discounts from the same function.[ Another thread](https://github.com/Shopify/function-examples/discussions/199) is tracking the inability to target line items.

**Reply by jacknavynicy** [reply]
Hi @dianacheptene so the scheduled release will have the fix for this discussion only right? for that another thread you mentioned is there any schedule for that to be released? Thank you!

**Reply by dianacheptene** [reply]
Hi jack,
I've replied in that thread.

**Comment by damir-samatov** [comment]
Right now Discount Extension Function is very limited in terms of it's capabilities and flexiblity, it only allows assigning a discount to single cart item (by variantID, which is confusing cuz it does not give control over specific line item I wish to discount). 
Although Function allows returning multiple targets with differrent discounts, it only applies the first one or the one with the highest discount amount, which does not make sense.

When can we expect those issues to be resolved?

Thank you!

**Reply by dianacheptene** [reply]
> Although Function allows returning multiple targets with differrent discounts, it only applies the first one or the one with the highest discount amount, which does not make sense.

Hi Damir, as mentioned above we're adding the ability to apply ALL applicable discounts in one function. We're a few weeks away from releasing this enhancement. 

>  by variantID, which is confusing cuz it does not give control over specific line item I wish to discount

We're tracking this issue in a separate [thread.](https://github.com/Shopify/function-examples/discussions/199)

**Comment by elvonkh** [comment]
@dianacheptene

Thank you for responding on this thread.
Just a clarification regarding the function usage.

Will application of discounts go with a single function? or should we create multiple functions?

Also this might not be related to this discussion, however discounts have limitation for amount of functions to be created per store.
The real issue that I have:

- I have created 5 automatic discounts in my app. I delete my app. I installed the app. I **cannot** create any discounts
- I have created 3 automatic discounts in my **first** application. I created 2 automatic discounts in **second** application. I cannot create 3rd discount in my **second** application as well as I cannot create 4th automatic discount in my **first** application

Probably, shopify has limitation for automatic discount PER SHOP. So the shop can bear 5 automatic discounts. Devs might be confused when developing a feature. Can you include this limitation in your documentation, please?

I cannot find any documentation regarding this. cc: https://community.shopify.com/c/shopify-functions/app-automatic-discount-limitation/td-p/1921328

**Reply by dianacheptene** [reply]
Noted. We'll update the docs to make it more clear.

**Comment by AndreaMorone** [comment]
Just to give an update on this topic. Seems that DISCOUNT APPLICATION STRATEGY::ALL is working and applying correctly, it's just the message that is wrong. From my tests it's like the message still uses DISCOUNT APPLICATION STRATEGY::FIRST.
<img width="1209" alt="screenshot" src="https://github.com/Shopify/function-examples/assets/7823934/8fc0a96c-4ac2-4122-bb50-b6c96029da95">
<img width="809" alt="Screenshot 2023-11-03 alle 18 05 45" src="https://github.com/Shopify/function-examples/assets/7823934/626163f8-fd7e-4509-a19a-3c9857f45c93">

**Reply by adamkitain** [reply]
Confirming we also see this!

**Reply by Vincent-Loi** [reply]
@AndreaMorone I get an error that `ALL` is not an acceptable enum value and using Shopify functions. How are you able to do this?

<img width="1027" alt="discount-strategy-error" src="https://github.com/Shopify/function-examples/assets/17518173/2407a826-8e35-4316-9f54-70e2e693c84e">

**Reply by AndreaMorone** [reply]
@Vincent-Loi you need to use the **unstable** api

**Reply by Vincent-Loi** [reply]
@AndreaMorone Ah might be a obvious question — where do I add / change to the unstable API?

**Reply by AndreaMorone** [reply]
In the extension toml

**Reply by X4mp** [reply]
How is it intended to be used with the generated `api.ts` file in the extension? Within this file the DiscountApplicationStrategy is defined with First and Maximum. I edited schema.graphql manually to add ALL as an option. But what's the actual correct way to update the local code to be able to use the unstable API options for the schema?
I also added unstable to the extesion.toml.

**Reply by nickwesselman** [reply]
See [this guide](https://shopify.dev/docs/apps/functions/input-output#api-versions) for changing API versions, you need to pull the new schema and regenerate types.

**Reply by joeainsworth** [reply]
Do we need to use "2024-01" for testing?

**Reply by AndreaMorone** [reply]
You need to use 'unstable'

**Reply by joeainsworth** [reply]
Thanks @AndreaMorone 
I change to `unstable` then ran `npm run shopify app function typegen` but does not seem to work (see screenshot).
I also tested function but error returned:
```
[
  {
    "path": [
      "discountApplicationStrategy"
    ],
    "explanation": "Expected value to not be null"
  }
]
```
Have I missed a step?
![Screenshot 2023-11-28 at 15 38 08](https://github.com/Shopify/function-examples/assets/5084107/8044bcdb-ada0-47b7-96de-744d9ba69daf)

**Reply by AndreaMorone** [reply]
Did you ran also https://shopify.dev/docs/apps/tools/cli/commands#function-schema?

**Reply by joeainsworth** [reply]
That's done it, thanks

**Reply by andriusvo** [reply]
I guess using unstable in production is not possible, am I right?

**Reply by rahultiwari-hulkapps** [reply]
I am also looking for this answer, can we use unstable in production? what are the consequences?

**Reply by AndreaMorone** [reply]
Yes, it’s possible. Many apps are already using it in production.

**Reply by rahultiwari-hulkapps** [reply]
Thank you, 
I was unable to find logs on the partner side, when using an unstable API? Is it true?

**Reply by AndreaMorone** [reply]
Yes, it works as any other function, it’s just a different API version that could return unexpected results/behavior.

**Reply by dostu** [reply]
The issue with wrong discount titles seems to be fixed 🥳

**Reply by AndreaMorone** [reply]
yep, I confirm! 
![screenshot-2023-12-04-17 15 32](https://github.com/Shopify/function-examples/assets/7823934/b02a91e1-66da-4a6e-ab6a-969fb82d3300)

![screenshot-2023-12-04-17 20 38](https://github.com/Shopify/function-examples/assets/7823934/f35611ff-5763-4f5f-a898-f61809bd845b)

**Reply by KapilAnncode** [reply]
Can you please tell me how I can pass "DiscountApplicationStrategy.All" in the function response, because in my api.ts file that's inside the "generated" subfolder of the created function. there is no option for all 

my api.ts file DiscountApplicationStrategy  contains like this
export enum DiscountApplicationStrategy {
  /** Only apply the first discount with conditions that are satisfied. */
  First = 'FIRST',
  /** Only apply the discount that offers the maximum reduction. */
  Maximum = 'MAXIMUM'
}

even after updating extension version to api_version = "2024-01"
and when I'm putting hardcoded  All= 'All', inside the DiscountApplicationStrategy  and when I'm running npm run shopify app function typegen command or run num run dev to remove hardcoded value from api.ts file DiscountApplicationStrategy  container.
![1](https://github.com/Shopify/function-examples/assets/108118307/046ed776-c0e4-4100-b082-c8a1da5019fa)
![2](https://github.com/Shopify/function-examples/assets/108118307/c18b291e-b3fc-46cc-a3e7-e6e54ab7cd56)
![3](https://github.com/Shopify/function-examples/assets/108118307/c4db56df-1ce6-4f4c-9bf9-117d7e45089c)

**Reply by AndreaMorone** [reply]
Did you try with
```javascript
npm run --silent shopify app function schema
```

**Reply by rgilbert82** [reply]
@KapilAnncode 
DiscountApplicationStrategy.All didn't work for me, but "ALL" did. All caps. Like this:
```
return {
      discounts: discountsArray,
      discountApplicationStrategy: "ALL"
};
```

**Reply by KapilAnncode** [reply]
Thanks, @AndreaMorone, Yes it worked for me after updating the function schema.

**Comment by killalau** [comment]
May I know if `DiscountApplicationStrategy::ALL` will be also available for order discount function or not?

**Reply by dianacheptene** [reply]
What use case are you looking to solve?

**Comment by dianacheptene** [comment]
All, the `DiscountApplicationStrategy: ALL` is now available in the [2024-01 release candidate](https://shopify.dev/docs/api/functions/reference/product-discounts/graphql/common-objects/discountapplicationstrategy?api%5Bversion%5D=2024-01). Please open new topics for any issues you may encounter.

**Reply by michelarteta** [reply]
Very nice, thank you @dianacheptene

**Comment by pbtripletree** [comment]
Seems like `DiscountStrategy.First` is used when a variant is targeted by two separate discounts. My expectation was that the discounts would stack, or at the very least `DiscountStrategy.Max` would be used in the event of overlap. We chalking this up as a feature or bug?

![Screenshot 2023-12-06 at 5 07 18 PM](https://github.com/Shopify/function-examples/assets/29215222/c5d9233f-3bff-4437-8472-1057b17e0a38)
![Screenshot 2023-12-06 at 5 07 00 PM](https://github.com/Shopify/function-examples/assets/29215222/920ef289-26f6-426b-8b7a-8cceca0fa938)

**Reply by meandillar** [reply]
It would be great to be able to choose the fallback, whether it be stacking, first or max!

**Reply by Mooh07** [reply]
I have a similar issue and it should be a bug.
Basically I have one automatic discount set up in admin dashboard that applies to all products, I have a function discount running in background with complex logic to apply discounts.

The automatic discount that applies to all products is 20%
Now I have 2 cart items, one that has a discount value of 50% applied through our discount function, the second item has a 5% discount.

Now the logically it should either stack or it should apply the greater applicable discount for second item, which makes it 20% instead of 5%, but the problem I'm getting 5%.

Screenshot for admin dashboard that indicates the bigger value should be taken in discounts, this is a bug
![image](https://github.com/Shopify/function-examples/assets/56565408/1cc4b27a-ef1b-444f-a51b-f2e41ade0617)

**Comment by ghost** [comment]
Gražutė

**Comment by nickwesselman** [comment]
Closing as this feature is complete and released.

**Comment by thosakwe** [comment]
Will `DiscountApplicationStrategy.ALL` be added to order discounts as well?

If not, then for anyone reading this, a workaround is to combine multiple discounts, into 1, where each eligible product variant is a target. The limitation here is that every product will receive the same discount.
